design/src/rfcs/rfc0003_presigning_api.md

```diff
@@ -0,0 +1,192 @@
+
+> Status: RFC
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+Several AWS services allow for pre-signed requests in URL form, which is described well by
+[S3's documentation on authenticating requests using query parameters](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html).
+
+This doc establishes the customer-facing API for creating these pre-signed URLs and how they will
+be implemented in a generic fashion in the SDK codegen.
+
+Terminology
+-----------
+
+To differentiate between the clients that are present in the generated SDK today, the following
+terms will be used throughout this doc:
+
+- **Smithy Client**: A `smithy_client::Client<C, M, R>` struct that is responsible for gluing together
+  the connector, middleware, and retry policy. This is not generated and lives in the `smithy-client` crate.
+- **Fluent Client**: A code-generated `Client<C, M, R>` that has methods for each service operation on it.
+  A fluent builder is generated alongside it to make construction easier.
+
+Pre-signed URL config
+---------------------
+
+Today, pre-signed URLs take an expiration time that's not part of the service API.
+The SDK will make this configurable as a separate struct so that there's no chance of name collisions, and so
+that additional fields can be added in the future. Fields added later will require defaulting for
+backwards compatibility.
+
+Since there's only one configuration field today, an `expires` function will be added to conveniently
+construct `PresigningConfig` without a builder. At the same time, a builder will be present for future-proofing:
+
+```rust
+#[non_exhaustive]
+#[derive(Debug, Clone)]
+pub struct PresigningConfig {
+    expires: Duration,
+}
+
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct Builder {
+  expires: Option<Duration>,
+}
+
+impl Builder {
+  pub fn expires(self, expires: Duration) -> Self { ... }
+  pub fn set_expires(self, expires: Option<Duration>) -> Self { ... }
+
+  pub fn build(self) -> PresigningConfig { ... }
+}
+
+impl PresigningConfig {
+    pub fn expires(expires: Duration) -> PresigningConfig {
+        Self::builder().expires(expires).build()
+    }
+
+    pub fn builder() -> Builder { ... }
+}
+```
+
+`PresigningConfig` is intended to be shared across AWS services, and will be passed by reference where it
+is taken as an argument.
+
+Fluent Pre-signed URL API
+-------------------------
+
+The generated fluent builders for operations that support pre-signing will have a `presigned()` method
+in addition to `send()` that will return a pre-signed URL rather than sending the request. For S3's GetObject,
+the usage of this will look as follows:
+
+```rust
+let config = aws_config::load_config_from_environment().await;
+let client = s3::Client::new(&config);
+let presigning_config = PresigningConfig::expires(Duration::from_secs(86400));
+let presigned_url: String = client.get_object()
```

Comment:
FWIW a general recommendation I've been given regarding presigned requests is to model the method and any associated headers rather than just the URL as a string.

Reply:
great catch, I totally missed this. Yes, `presign` should return an `http::Request<()>` or similar, not a string

Reply:
I would even suggest returning a `PresignedRequest` type (naming tbd). It can contain the `http::Request<()>` value internally and delegate methods as appropriate.